Add support for conditional DETR model

DeppMeng opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Added codes and documentations for conditioonal DETR model. The conditional DETR files are created by using the "add-new-model-like" feature of CookieCutter, based on DETR codes. All tests are passed. One thing I want to ask is that, I have converted the pretrained weights, how shoud I give these weights to you?


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/Atten4Vis/ConditionalDETR/issues/21
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

DeppMeng committed:
Added conditional DETR (#1)

* added conditional detr

* added conditional detr source

* added first part of conditional detr

* added conditional_detr

Co-authored-by: Depu Meng <depumeng@Depus-MacBook-Pro.local>

DeppMeng committed:
Added conditional DETR (#1)

* added conditional detr

* added conditional detr source

* added first part of conditional detr

* added conditional_detr

Co-authored-by: Depu Meng <depumeng@Depus-MacBook-Pro.local>

Narsil committed:
[BC] Fixing usage of text pairs (#17324)

* [BC] Fixing usage of text pairs

The BC is actually preventing users from misusing the pipeline since
users could have been willing to send text pairs and the pipeline would
instead understand the thing as a batch returning bogus results.

The correct usage of text pairs is preserved in this PR even when that
makes the code clunky.

Adds support for {"text":..,, "text_pair": ...} inputs for both dataset
iteration and more explicit usage to pairs.

* Updating the doc.

* Update src/transformers/pipelines/text_classification.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update src/transformers/pipelines/text_classification.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update tests/pipelines/test_pipelines_text_classification.py

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

* quality.

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

NielsRogge commented on docs/source/en/model_doc/conditional_detr.mdx:
```diff
@@ -0,0 +1,66 @@
+<!--Copyright 2021 The HuggingFace Team. All rights reserved.
```
```suggestion
<!--Copyright 2022 The HuggingFace Team. All rights reserved.
```

NielsRogge commented on docs/source/en/model_doc/conditional_detr.mdx:
```diff
@@ -0,0 +1,66 @@
+
+## Overview
+
+The conditional_detr model was proposed in [ConditionalDETR](https://arxiv.org/abs/2108.06152) by Depu Meng, Xiaokang Chen, Zejia Fan, Gang Zeng, Houqiang Li, Yuhui Yuan, Lei Sun, Jingdong Wang.
```
```suggestion
The Conditional DETR model was proposed in [ConditionalDETR](https://arxiv.org/abs/2108.06152) by Depu Meng, Xiaokang Chen, Zejia Fan, Gang Zeng, Houqiang Li, Yuhui Yuan, Lei Sun, Jingdong Wang.
```
We usually also add a one-line summary here of the model, after this sentence.

DeppMeng replied:
Solved, also added a one-line summary.

NielsRogge commented on src/transformers/models/conditional_detr/configuration_conditional_detr.py:
```diff
@@ -0,0 +1,204 @@
+# coding=utf-8
+# Copyright 2021 Facebook AI Research and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```

NielsRogge commented on src/transformers/models/conditional_detr/configuration_conditional_detr.py:
```diff
@@ -0,0 +1,204 @@
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+
```
```suggestion
```

NielsRogge commented on src/transformers/models/conditional_detr/convert_conditional_detr_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,290 @@
+# coding=utf-8
+# Copyright 2020 The HuggingFace Inc. team.
```
```suggestion
# Copyright 2022 The HuggingFace Inc. team.
```

NielsRogge commented on src/transformers/models/conditional_detr/convert_conditional_detr_original_pytorch_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,290 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Convert CONDITIONAL_DETR checkpoints."""
```
```suggestion
"""Convert Conditional DETR checkpoints from the original repository.

URL: https://github.com/Atten4Vis/ConditionalDETR
"""
```

NielsRogge commented on src/transformers/models/conditional_detr/feature_extraction_conditional_detr.py:
```diff
@@ -0,0 +1,935 @@
+    return color
+
+
+class ConditionalDETRFeatureExtractor(FeatureExtractionMixin, ImageFeatureExtractionMixin):
```
I'm not sure whether there's a need for a new feature extractor, as this one seems identical to DetrFeatureExtractor.

In that case, we recommend users to use the AutoFeatureExtractor class, which will instantiate a DetrFeatureExtractor behind the scenes, based on the checkpoint provided:

```
from transformers import AutoFeatureExtractor

feature_extractor = AutoFeatureExtractor.from_pretrained("Attn4Vis/conditional-detr")
```
You just need to add ("conditional_detr", "DetrFeatureExtractor") to models/auto/feature_extraction_auto.py

DeppMeng replied:
Actually the FeatureExtractor is not identical to DETR. The post_process function in conditional DETR is different to DETR. So I think we do need a separate FeatureExtractor

NielsRogge commented on src/transformers/models/conditional_detr/modeling_conditional_detr.py:
```diff
@@ -0,0 +1,2556 @@
+# coding=utf-8
+# Copyright 2021 Facebook AI Research The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 Attn4Vis and The HuggingFace Inc. team. All rights reserved.
```
You might add another author here

NielsRogge commented on src/transformers/models/conditional_detr/modeling_conditional_detr.py:
```diff
@@ -0,0 +1,2556 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch CONDITIONAL_DETR model."""
```
```suggestion
""" PyTorch Conditional DETR model."""
```

NielsRogge commented on src/transformers/models/conditional_detr/modeling_conditional_detr.py:
```diff
@@ -0,0 +1,2556 @@
+
+# BELOW: utilities copied from
+# https://github.com/facebookresearch/detr/blob/master/backbone.py
+# Copied from transformers.models.detr.modeling_detr.DetrFrozenBatchNorm2d with Detr->ConditionalDETR
```
Very nice to leverage Copied from statements (running `make fix-copies` locally ensures the consistency of the code)

NielsRogge commented on src/transformers/models/conditional_detr/modeling_conditional_detr.py:
```diff
@@ -0,0 +1,2556 @@
+        return out, pos
+
+
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
```
Here, a Copied from statement can probably also be added

NielsRogge commented on src/transformers/models/conditional_detr/modeling_conditional_detr.py:
```diff
@@ -0,0 +1,2556 @@
+    def _v_shape(self, tensor: torch.Tensor, seq_len: int, bsz: int):
+        return tensor.view(bsz, seq_len, self.num_heads, self.v_head_dim).transpose(1, 2).contiguous()
+
+    # def with_pos_embed(self, tensor: torch.Tensor, position_embeddings: Optional[Tensor]):
+    #     return tensor if position_embeddings is None else tensor + position_embeddings
```
To be removed?

NielsRogge commented on docs/source/en/model_doc/conditional_detr.mdx:
```diff
@@ -0,0 +1,66 @@
+
+[[autodoc]] ConditionalDETRConfig
+
+## ConditionalDETRFeatureExtractor
+
+[[autodoc]] ConditionalDETRFeatureExtractor
+    - __call__
+    - pad_and_create_pixel_mask
+    - post_process
+    - post_process_segmentation
+    - post_process_panoptic
```
There's probably no need for a new feature extractor, see comments below

DeppMeng replied:
Check my reply to the other comment

Narsil committed:
Adding `batch_size` test to QA pipeline. (#17330)

patrickvonplaten committed:
[OPT] Run test in lower precision on GPU (#17353)

* [OPT] Run test only in half precision

* up

* up

* up

* up

* finish

* fix on GPU

* Update tests/models/opt/test_modeling_opt.py

patrickvonplaten committed:
[Generation] Fix Transition probs (#17311)

* [Draft] fix transition probs

* up

* up

* up

* make it work

* fix

* finish

* update

nadahlberg committed:
fix for 17292 (#17293)

ddobokki committed:
Fix bug in Wav2Vec2 pretrain example (#17326)

patrickvonplaten committed:
[Test OPT] Add batch generation test opt (#17359)

* up

* up

sgugger committed:
Pin dill to fix examples (#17368)

* Pin dill for now

* Try this version?

* force install

* Actually use dep in testing

* Try a larger pin

stancld committed:
Fix a typo relative_postion_if_large -> relative_position_if_large (#17366)

loubnabnl committed:
Fix CodeParrot training script (#17291)

* average loss over batches and accumulated steps for tracking

* fix layernorm weight decay

* use AdamW from Pytorch instead of Transformers

* add shuffling of sequences inside the batches

* add shuffling of sequences inside the batches

* add logging dir and reformat code

* fix lr tracking

* remove Mistral scaling

* keep Mistral scaling

* reformat code

* fix error

* fix error

* use shuffling function from Pytorch

* remove argument for shuffling batch sequences as it isn't optional

* update package versions and install accelerate from source

* remove unused package

* Update loss average over accumulated steps

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* Update loss average over accumulated steps

Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

* use one shuffle buffer argument

* compute avg_loss in one line

Co-authored-by: Loubna ben allal <loubnabenallal@gmail.com>
Co-authored-by: Leandro von Werra <lvwerra@users.noreply.github.com>

garyhlai committed:
Correct & Improve Doctests for LayoutLMv2 (#17168)

* add inference example to LayoutLMv2ForQuestionAnswering, passing doctest

* add loss example to LayoutLMv2ForQuestionAnswering, passing doctest

* Add correct doctest for LayoutLMv2ForTokenClassification, passing doctest

* add correct doctest for LayoutLMv2ForSequenceClassification, passing test

* add correct doctest for LayoutLMv2Model, passing test

* make fixup

* fix to address review comments

* make style

* fix doctest line break issue, add to documentaiton_tests.txt, address review comments

* move comment about layoutlmv2 dependencies to the doc page

* format doc page as suggested

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* delete extraneous backtick

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

AnugunjNaman committed:
Fix cvt docstrings (#17367)

mxschmdt committed:
Fix Comet ML integration (#17381)

Callback function `on_train_end` crashed if Comet ML integration was
used but `COMET_MODE` set to `DISABLE`

michaelbenayoun committed:
Traced models serialization and torchscripting fix (#17206)

* Fix torch.jit.script and pickling issues

* Fix get_attr issues

* Fix import in function

* Fix GPT-J and T5 tracing for torch=1.11

* Gate graph surgery on torch version

* Modeling minor changes to enable TorchScripting

* Model serialization / deserialization test

* Remove _assert_is_none users

sgugger committed:
Use Accelerate in `from_pretrained` for big model inference (#17341)

* Initial work

* More or less finished with first draft

* Update src/transformers/modeling_utils.py

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Update src/transformers/modeling_utils.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

* Fix randomly initialized weights

* Update src/transformers/modeling_utils.py

Co-authored-by: Lysandre Debut <lysandre.debut@reseau.eseo.fr>

* Address review comments

* Rename DeepSpeed folder to temporarily fix the test issue?

* Revert to try if Accelerate fix works

* Use latest Accelerate release

* Quality and fixes

* Style

* Quality

* Add doc

* Test + fix

* More blocks

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>
Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>
Co-authored-by: Lysandre Debut <lysandre.debut@reseau.eseo.fr>

vfbd committed:
OPTForCausalLM lm_head input size should be config.word_embed_proj_dim (#17225)

sgugger committed:
Add support for `device_map="auto"` to OPT (#17382)

DeppMeng committed:
rebased from main

DeppMeng committed:
Update docs/source/en/model_doc/conditional_detr.mdx

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

DeppMeng committed:
Update docs/source/en/model_doc/conditional_detr.mdx

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

DeppMeng committed:
solved some documentation and undeleted comments issues posted in PR comments

DeppMeng commented on docs/source/en/model_doc/conditional_detr.mdx:
```diff
@@ -0,0 +1,66 @@
+head is able to attend to a band containing a distinct region, e.g., one object extremity or a region inside the object box. This narrows down the spatial range for localizing the distinct regions for object classification and box
+regression, thus relaxing the dependence on the content embeddings and easing the training. Empirical results show
+that conditional DETR converges 6.7× faster for the backbones R50 and R101 and 10× faster for stronger backbones
+DC5-R50 and DC5-R101. Code is available at https://github.com/Atten4Vis/ConditionalDETR.*
```
Cleaned the formatting

DeppMeng committed:
fixed error caused by unsolved conflicts
